cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,381 @@
+ * limitations under the License.
+ */
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/scatter.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/strings/detail/scatter.cuh>
+#include <cudf/strings/string_view.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+template <typename T, typename MapIterator>
+rmm::device_vector<T> make_gather_map(MapIterator scatter_iter,
+    size_type scatter_rows, size_type gather_rows, cudaStream_t stream)
+{
+  static_assert(std::is_signed<T>::value,
+    "Need different invalid index if unsigned index types are added");
+  auto const invalid_index = static_cast<T>(-1);
+
+  // Convert scatter map to a gather map
+  auto gather_map = rmm::device_vector<T>(gather_rows, invalid_index);
+  thrust::scatter(rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<T>(0),
+    thrust::make_counting_iterator<T>(scatter_rows),
+    scatter_iter, gather_map.begin());
+
+  return gather_map;
+}
+
+template <typename MapIterator>
+void gather_bitmask(table_view const& source, MapIterator gather_map,
+    std::vector<std::unique_ptr<column>>& target,
+    rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  // Create null mask if source is nullable but target is not
+  for (size_t i = 0; i < target.size(); ++i) {
+    if (source.column(i).nullable() and not target[i]->nullable()) {
+      auto mask = create_null_mask(target.size(), mask_state::ALL_VALID, stream, mr);
+      target[i]->set_null_mask(std::move(mask), 0);
+    }
+  }
+
+  auto const device_source = table_device_view::create(source, stream);
+
+  // Make device array of target bitmask pointers
+  thrust::host_vector<bitmask_type*> target_masks(target.size());
+  std::transform(target.begin(), target.end(), target_masks.begin(),
+    [](auto const& col) { return col->mutable_view().null_mask(); });
+  rmm::device_vector<bitmask_type*> d_target_masks(target_masks);
+  auto target_rows = target.front()->size();
+
+  // Compute block size
+  auto bitmask_kernel = gather_bitmask_kernel<true, decltype(gather_map)>;
+  constexpr size_type block_size = 256;
+  size_type const grid_size = grid_1d(target_rows, block_size).num_blocks;
+
+  auto d_valid_counts = rmm::device_vector<size_type>(target.size());
+  bitmask_kernel<<<grid_size, block_size, 0, stream>>>(*device_source,
+    gather_map, d_target_masks.data().get(), target_rows, d_valid_counts.data().get());
+
+  // Copy the valid counts into each column
+  auto valid_counts = thrust::host_vector<size_type>(d_valid_counts);
+  auto begin = thrust::make_tuple(target.begin(), valid_counts.begin());
+  auto end = thrust::make_tuple(target.end(), valid_counts.end());
+  thrust::for_each(thrust::host,
```

Comment:
I don't think it's worth it to use thrust here. This is actually more obfuscated and more code than a for loop without the zip iterators, IMO. Plus it's host-only. There may be a way to use a std::transform (binary op version) which would avoid need for a zip iterator, but even there a for loop may be simpler.

Reply:
Yes, a for_each that takes two iterators would be perfect... I considered transform since it does accept two iterators, but it seems you're not supposed to modify the values and I'd also need a dummy output iterator to discard the unused output.

Reply:
Ok, I used transform. thrust::make_discard_iterator() to the rescue.

Reply:
Replaced transform with range for